File: December2021/NthMagicalNumber.java
class Solution {
	public int nthMagicalNumber(int n, int a, int b) {

		long A = a, B = b;
		long mod = (long)(Math.pow(10,9)+7);


		long left = Math.min(a,b), right = (long)n*Math.min(a,b);

		// Finding GCD
		while (B > 0) {
			long tmp = A;
			A = B;
			B = tmp % B;
		}

		long lcm = (a*b)/A;     //(n1*n2)/GCD

		while (left < right) {
			long m = left+(right-left)/2;

			if ((m / a) + (m / b) - (m / lcm) < n) left = m + 1;

			else right = m;
		}
		return (int)(left % mod);
	}
}

File: December2021/C++/Nth_Magical_Number.cpp
class Solution {
    int gcd(int a,int b)
    {
        if(a==0) return b;
        return gcd(b%a,a);
    }
public:
    int nthMagicalNumber(int n, int a, int b) {
        long long  l=min(a,b);
        long long  r= l*n;
        long long lcm=(a*b)/gcd(a,b);
        long long mid,factor;
        while(l<r)
        {
            mid=l+(r-l)/2;
            factor=mid/a +mid/b - mid/lcm;
            if(factor<n) l=mid+1;
            else r=mid;
        }
        int mod=1e9+7;
        return l%mod;
    }
};